pkg/liveshare/ports.go

```diff
@@ -0,0 +1,134 @@
+package liveshare
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/sourcegraph/jsonrpc2"
+	"golang.org/x/sync/errgroup"
+)
+
+// Port describes a port exposed by the container.
+type Port struct {
+	SourcePort                       int    `json:"sourcePort"`
+	DestinationPort                  int    `json:"destinationPort"`
+	SessionName                      string `json:"sessionName"`
+	StreamName                       string `json:"streamName"`
+	StreamCondition                  string `json:"streamCondition"`
+	BrowseURL                        string `json:"browseUrl"`
+	IsPublic                         bool   `json:"isPublic"`
+	IsTCPServerConnectionEstablished bool   `json:"isTCPServerConnectionEstablished"`
+	HasTLSHandshakePassed            bool   `json:"hasTLSHandshakePassed"`
+	Privacy                          string `json:"privacy"`
+}
+
+type PortChangeKind string
+
+const (
+	PortChangeKindStart  PortChangeKind = "start"
+	PortChangeKindUpdate PortChangeKind = "update"
+)
+
+type PortUpdate struct {
```

Comment:
I merged the two in the latest commit.

In this case, the StatusCode and ErrorDetail are coming from the event that's being received through RPC from the Codespace. These contain information about why the port may have failed forwarding either in the Codespace or the service. These are useful because they let us disambiguate between genuine errors vs requests that were blocked by the service due to being out of policy, since we want to show different error messages in each case. Success in this case helps us disambiguate between the two possible events that can be sent by the Codespace (SharingSucceeded vs SharingFailed). The error that's returned by the function catches the case where unmarshalling returns an error. I agree there's a lot that could go wrong here, so I've added extra comments to explain which properties are sent by the event.